### DISCUSSION TITLE
Add sub-categories in a table with an option to change them

### DISCUSSION THREAD

**Discussion by BismaIftekhar** [Budibase/budibase#10664] [Help] [answered]
Hello everyone,

I want to create a table with a certain row having sub-categories. The issue is they are not fixed and the user should have an option of changing these sub-categories. Is this possible? Please find below, for example the teachers should be able to change the courses and the associated percentages.

![Budibase-example](https://github.com/Budibase/budibase/assets/96243721/d2b6aadd-359c-4044-9343-4a809d941aaa)

Can someone please guide me what to do?

**Comment by melohagan** [comment, marked as answer]
Hey @BismaIftekhar 

This is an interesting use-case - thank you for the diagram!

It was fairly complicated, but I was able to come up with a solution using the community [Super Table](https://github.com/poirazis/bb-component-SuperTable) by @poirazis in combination with the Budibase [JSON](https://docs.budibase.com/docs/json) data type.

---


### Step 1 - Create the Data table

Create a new table called **Scores** to keep track of the marks for each attribute, i.e. Attendance, Behaviour, Courses.

This tutorial won't cover [relationships](https://docs.budibase.com/docs/relationships), but most likely you would want to link this table to your **Teachers** table. 

Create a JSON type column called **Attributes**. Click `Open schema editor` and provide the following JSON schema:

```json
{
  "name": "Courses",
  "subcat1": "Math",
  "submarks1": 20,
  "subcat2": "English",
  "submarks2": 30,
  "subcat3": "Language",
  "submarks3": 20,
  "subcat4": "Technical Skills",
  "submarks4": 10,
  "marks": 100,
  "hasSubCategories": true
}
```

The *Form* tab should appear as follows:
![Screenshot 2023-06-01 at 15 41 01](https://github.com/Budibase/budibase/assets/101575380/3ba13e95-286d-47bd-8948-972a74280a5f)

Make sure to `Save Column`. (You'll need to click it twice). 

You will notice from the schema that we have hard-coded four sub-categories. This is a bit of a limitation in that you need to know the number of attributes ahead of time. There is an open feature request for handling dynamic form inputs: https://github.com/Budibase/budibase/issues/5658

Also note the *hasSubCategories* property. This will determine if the sub-category fields should be shown or not.

![Screenshot 2023-06-01 at 15 44 46](https://github.com/Budibase/budibase/assets/101575380/4e48f6d9-4054-4558-8694-570590c65b60)

### Step 2 - Adding the Super Table

Make sure you have imported the [SuperTable](https://github.com/poirazis/bb-component-SuperTable) plugin as well as the companion **Super Column** and **Super Cell** plugins. 

More information can be found in the [custom plugins](https://docs.budibase.com/docs/custom-plugin) docs.

Next add the necessary components to your screen:

1. Data provider (for the Scores table)
2. Nest the Super Table - **make sure the ID Column is set to _id**
![Screenshot 2023-06-01 at 15 50 33](https://github.com/Budibase/budibase/assets/101575380/4500e58b-73ae-4820-8408-9e12d87ecdae)

3. Nest three Super Columns. One for the name of the attribute, one for any sub-categories, and one for the marks.
4. Nest a Super Cell within the attribute name column. Provide the *Value* binding: `{{ Attributes Column.Cell Value.name }}`
5. Nest a Super Cell within the sub category column. Leave the *Value* field blank.
6. Nest a Super Cell within the marks column. Again leave the *Value* field blank.

### Step 3 - Adding the form fields

With the table structure setup, we need to add in the form fields to allow the sub categories and marks to be edited. 

Within the sub category column component, nest a Form component and select the **Scores** table as the *Schema*.
Next nest a Container component, and within it a field group component and add four Text field (one for each sub category). 
Also add an Icon component to the container for saving the form (we'll implement the click action later).

Assign a field for each from the dropdown, for example:
![Screenshot 2023-06-01 at 15 57 16](https://github.com/Budibase/budibase/assets/101575380/ef4b6e04-6bb9-47cf-84ec-7db68b0efe72)

Also assign the correct default value. For example: `{{ Sub Category Column.Cell Value.subcat1 }}`

Another step we need to do is only showing the field group and save icon if *hasSubCategories* is set. We do this through [Conditional UI](https://docs.budibase.com/docs/conditions).

Click on the container component, and select the *Conditions* tab from the settings panel. Click `Configure conditions` and setup the following condition:
![Screenshot 2023-06-01 at 16 03 40](https://github.com/Budibase/budibase/assets/101575380/992b976c-f56b-41fb-8637-9f964b62d12e)

With the sub categories done, follow similar steps for the marks form fields. There are a couple of differences: use *Number* fields instead of text fields, and within the container have one field called *Marks*, and below a field group for the sub-category marks. 
The conditional UI will be the same as above, but apply it to the field group instead. The *Marks* field will also need the condition, but inverse the boolean flag to false.

At the end your component tree should look similar to the following:
![Screenshot 2023-06-01 at 16 08 46](https://github.com/Budibase/budibase/assets/101575380/5478a90d-5593-4b21-90d5-fd0555251b4f)

### Step 4 - Saving the fields

The final step is to make the save icons update the existing JSON with the newly entered data. Click on the Save icon component for the sub-categories form and click `Define actions`.

Use the `Add Column` button to add the **_id** field (indicate which row should be updated) as well as the Attributes JSON field:
![Screenshot 2023-06-01 at 16 11 35](https://github.com/Budibase/budibase/assets/101575380/d7aafbe0-174c-4cb2-a46a-c09bc5072750)

The JavaScript binding is as follows:

```js
return {
	...$("Sub Category Column.Cell Value"),
	subcat1: $("Sub Form.Fields.Attributes.subcat1"),
	subcat2: $("Sub Form.Fields.Attributes.subcat2"),
	subcat3: $("Sub Form.Fields.Attributes.subcat3"),
	subcat4: $("Sub Form.Fields.Attributes.subcat4"),
}
```

Here we are taking the existing attributes data and updating the subcat 1-4 fields with the form fields.
Make sure to `Save`!

The **Save Row** action for the Marks form will be very similar, just different form fields:

![Screenshot 2023-06-01 at 16 13 44](https://github.com/Budibase/budibase/assets/101575380/de687bfe-2102-4012-bd21-d51283ab40dc)

JavaScript binding:

```js
let cellVal = $("Marks Column.Cell Value")

if (cellVal.hasSubCategories) {
	return {
		...cellVal,
		submarks1: $("Marks Form.Fields.Attributes.submarks1"),
		submarks2: $("Marks Form.Fields.Attributes.submarks2"),
		submarks3: $("Marks Form.Fields.Attributes.submarks3"),
		submarks4: $("Marks Form.Fields.Attributes.submarks4"),
	}
}

return {
	...cellVal,
	marks: $("Marks Form.Fields.Attributes.marks")
}
```
 
There's a little but of logic here to determine if the marks are for sub-categories or the main category.

### Result

![matrix](https://github.com/Budibase/budibase/assets/101575380/eb3b813f-f35c-4bc9-b673-477b14f6a7e8)

**App export**: 
[Super Table Matrix-export-1685629864687.tar.gz](https://github.com/Budibase/budibase/files/11626881/Super.Table.Matrix-export-1685629864687.tar.gz)

**Reply by BismaIftekhar** [reply]
Apologies for the late reply. Firstly, thank you so much @melohagan for such an elaborate response. Did not have the time to test it yet but the results seems like what I am aiming for. 
Thank you. again :)

**Comment by BismaIftekhar** [comment]
Hello @melohagan another question please if you could help.

"How to retrieve data from a public form with dynamic entries?"


![Budibase_Publicform_issue](https://github.com/Budibase/budibase/assets/96243721/cbf61fbe-61ff-4949-9708-3bbb873bc85f)


So,

Step 1: 
The teacher creates entries in a Budibase Table that are dependant on classses for example. So Teachers of 2 Blue can create their own assessment criterias such as arithmetic, algebra etc.

Step2:
The students of 2Blue log in and grade themselves in those.

Question:
How to retrieve data in Budibase (the criterias and their marks) from the filled form by students considering they are not constant?

Once that data is written in the table, teachers should then be able to grade them too.

Any help would be appreciated. 
Thank you
